### PR TITLE
Issue #374: Fix Permitted directories bug

### LINKED ISSUE #374: Permitted dirs must not evaluate to include empty strings
Apparently exploding an empty string gives an array of one empty string.

### PR DESCRIPTION
**Pull request by jaypha**
Resolves #374 

Add checks to get_permitted_dirs() to exclude empty values.
Fix unit tests.
Add more unit tests
Remove empty lines before saving value.
Moved admin_setting_permitted_directories to classes/admin.

### CODE REVIEW

**Comment by keevan** [comment]
Also, on a loosely related note, are you able to move the admin setting under the directory `classes/admin/*` as per https://github.com/catalyst/moodle-tool_dataflows/issues/145

Related: 
https://docs.moodle.org/dev/Coding_style#Namespaces
https://docs.moodle.org/dev/Core_APIs#Admin_settings_API_.28admin.29

File: tests/tool_dataflows_helper_test.php [hunk 1]
```diff
@@ -37,15 +37,42 @@ protected function setUp(): void {
...
+                '/* A comment' . PHP_EOL . ' over two lines */' . PHP_EOL . '/tmp' . PHP_EOL . '/var',
+                ['/tmp', '/var']
+            ],
+            [
+                '# comment' . PHP_EOL . '/var/tmp   # more comments.',
+                ['/var/tmp']
+            ],
+            [
+                '/var/[dataroot]/tmp',
+                ['/var/[dataroot]/tmp']
```
Comment (keevan) [inline tests/tool_dataflows_helper_test.php:74]: I'm on the fence about whether or not this should be added as a multiline string, purely for readability. Reading the PHP_EOL and mentally pushing it onto the next line is a bit painful, and you cannot glance at the test to see if there's a case it's not yet covering.

Would you be keen to adjust this? I think it should also be expected that there might be some spacing before / after each line and it should be handled gracefully, and so you should be able to get what you want with the indenation.

e.g.
```
["
    /home/me/tmp
    [dataroot]/tmp
    ...other
    # a comment
", [.. valid options here]]
```

File: lang/en/tool_dataflows.php [hunk 2]
```diff
@@ -37,7 +37,7 @@
 $string['permitted_dirs'] = 'Permitted directories';
 $string['permitted_dirs_desc'] = "List directories here to allow them to be read from/written to by dataflow steps.
     One directory per line. Each directory must be an absolute path. You can use the place holder '{\$a}' for the
-    site's data root directory.\n
+    site's data root directory. You can also include comments using /* and */.\n
```
Comment (keevan) [inline lang/en/tool_dataflows.php:40]: Should also mention `#` based comments

Comment (brendanheywood) [reply lang/en/tool_dataflows.php:40]: +1 but let’s keep moving too

File: tests/tool_dataflows_helper_test.php (hunk 1 above)
Comment (jaypha) [reply tests/tool_dataflows_helper_test.php:74]: Do you mean replacing PHP_EOL with hard newlines?
As for spacing before/after, the lines get trimmed before processing.

File: lang/en/tool_dataflows.php (hunk 2 above)
Comment (jaypha) [reply lang/en/tool_dataflows.php:40]: I'll submit another PR as a followup.

Comment (jaypha) [reply lang/en/tool_dataflows.php:40]: This is being fixed as  a part of PR #417.

**Pull request merged**